src/gpuarray_buffer_cuda.c

```diff
@@ -456,17 +456,15 @@ static int cuda_read(void *dst, gpudata *src, size_t srcoff, size_t sz) {
 
     cuda_enter(ctx);
 
-    ctx->err = cuEventSynchronize(src->ev);
-    if (ctx->err != CUDA_SUCCESS) {
-      cuda_exit(ctx);
-      return GA_IMPL_ERROR;
-    }
+    cuda_wait(src, CUDA_WAIT_READ);
 
-    ctx->err = cuMemcpyDtoH(dst, src->ptr + srcoff, sz);
+    ctx->err = cuMemcpyDtoHAsync(dst, src->ptr + srcoff, sz, ctx->s);
```

Comment:
Why this is safe from Theano point of view? Do the host have a complete copy after that? If not, then other ops on the CPU could start to compute on it before the transfer is complete.